clab/docker.go

```diff
@@ -179,6 +187,52 @@ func (c *cLab) CreateContainer(ctx context.Context, node *Node) (err error) {
 	return createContainerNS(node.Pid, node.LongName)
 }
 
+func (c *cLab) PullImageIfRequired(ctx context.Context, node *Node) (err error) {
+	filter := filters.NewArgs()
+	filter.Add("reference", node.Image)
+
+	ilo := types.ImageListOptions{
+		All:     false,
+		Filters: filter,
+	}
+
+	log.Debugf("Looking up %s Docker image", node.Image)
+
+	images, err := c.DockerClient.ImageList(ctx, ilo)
+	if err != nil {
+		return err
+	}
+
+	// Image doesn't exist, need to pull it
+	if len(images) == 0 {
+		// might need canonical name e.g.
+		//    -> alpine == docker.io/library/alpine
+		//    -> foo/bar == docker.io/foo/bar
+		//    -> docker.elastic.co/elasticsearch/elasticsearch == docker.elastic.co/elasticsearch/elasticsearch
+		canonicalImageName := node.Image
+		slashCount := strings.Count(node.Image, "/")
+
+		switch slashCount {
+		case 0:
+			canonicalImageName = "docker.io/library/" + node.Image
+		case 1:
+			canonicalImageName = "docker.io/" + node.Image
+		}
+
+		log.Infof("Pulling %s Docker image", node.Image)
+		reader, err := c.DockerClient.ImagePull(ctx, canonicalImageName, types.ImagePullOptions{})
+		// must read from reader, otherwise image is not properly pulled
+		io.Copy(ioutil.Discard, reader)
```

Comment:
I also think maybe you should read to `os.Stdout`? I think that way you will be able to show the progress of image pulling to a user during the lab instantiation

Reply:
Agree, but after having seen the output I've decided to discard.

And that's just the output for the tiny alpine image...
```
Start -  docker.io/library/alpine
{"status":"Pulling from library/alpine","id":"latest"}

{"status":"Pulling fs layer","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":29170,"total":2797541},"progress":"[\u003e                                                  ]  29.17kB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":425216,"total":2797541},"progress":"[=======\u003e                                           ]  425.2kB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":781568,"total":2797541},"progress":"[=============\u003e                                     ]  781.6kB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":1207552,"total":2797541},"progress":"[=====================\u003e                             ]  1.208MB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":1604061,"total":2797541},"progress":"[============================\u003e                      ]  1.604MB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":2071005,"total":2797541},"progress":"[=====================================\u003e             ]  2.071MB/2.798MB","id":"df20fa9351a1"}

{"status":"Downloading","progressDetail":{"current":2464221,"total":2797541},"progress":"[============================================\u003e      ]  2.464MB/2.798MB","id":"df20fa9351a1"}

{"status":"Verifying Checksum","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Download complete","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Extracting","progressDetail":{"current":32768,"total":2797541},"progress":"[\u003e                                                  ]  32.77kB/2.798MB","id":"df20fa9351a1"}

{"status":"Extracting","progressDetail":{"current":2064384,"total":2797541},"progress":"[====================================\u003e              ]  2.064MB/2.798MB","id":"df20fa9351a1"}

{"status":"Extracting","progressDetail":{"current":2797541,"total":2797541},"progress":"[==================================================\u003e]  2.798MB/2.798MB","id":"df20fa9351a1"}

{"status":"Pull complete","progressDetail":{},"id":"df20fa9351a1"}

{"status":"Digest: sha256:185518070891758909c9f839cf4ca393ee977ac378609f700f60a771a2dfe321"}

{"status":"Status: Downloaded newer image for alpine:latest"}

Done -  docker.io/library/alpine
```

Reply:
valid point. especially considering that we will pull in parallel quite soon.